crates/bevy_ecs/src/schedule_v3/executor/multi_threaded.rs

```diff
@@ -0,0 +1,498 @@
+        let task = async move {
+            #[cfg(feature = "trace")]
+            let system_guard = system_span.enter();
+            // SAFETY: access is compatible
+            unsafe { system.run_unsafe((), world) };
+            #[cfg(feature = "trace")]
+            drop(system_guard);
+            sender
+                .send(system_index)
+                .await
+                .unwrap_or_else(|error| unreachable!("{}", error));
+        };
+
+        #[cfg(feature = "trace")]
+        let task = task.instrument(task_span);
+
+        let system_meta = &self.system_task_metadata[system_index];
+        self.active_access
+            .extend(&system_meta.archetype_component_access);
+
+        if system_meta.is_send {
+            scope.spawn(task);
+        } else {
+            self.local_thread_running = true;
+            scope.spawn_on_scope(task);
+        }
+
+        self.ready_systems.set(system_index, false);
+        self.running_systems.insert(system_index);
+    }
+
+    fn spawn_exclusive_system_task<'scope>(
+        &mut self,
+        scope: &Scope<'_, 'scope, ()>,
+        system_index: usize,
+        schedule: &'scope SystemSchedule,
+        world: &'scope mut World,
+    ) {
+        // SAFETY: system was not already running
+        let system = unsafe { &mut *schedule.systems[system_index].as_ptr() };
+
+        #[cfg(feature = "trace")]
+        let task_span = info_span!("system_task", name = &*system.name());
+        #[cfg(feature = "trace")]
+        let system_span = info_span!("system", name = &*system.name());
+
+        let sender = self.sender.clone();
+        if is_apply_system_buffers(system) {
+            // TODO: avoid allocation
+            let mut unapplied_systems = self.unapplied_systems.clone();
+            let task = async move {
+                #[cfg(feature = "trace")]
+                let system_guard = system_span.enter();
+                Self::apply_system_buffers(&mut unapplied_systems, schedule, world);
+                #[cfg(feature = "trace")]
+                drop(system_guard);
+                sender
+                    .send(system_index)
+                    .await
+                    .unwrap_or_else(|error| unreachable!("{}", error));
+            };
+
+            #[cfg(feature = "trace")]
+            let task = task.instrument(task_span);
+            scope.spawn_on_scope(task);
+        } else {
+            let task = async move {
+                #[cfg(feature = "trace")]
+                let system_guard = system_span.enter();
+                system.run((), world);
+                #[cfg(feature = "trace")]
+                drop(system_guard);
+                sender
+                    .send(system_index)
+                    .await
+                    .unwrap_or_else(|error| unreachable!("{}", error));
+            };
+
+            #[cfg(feature = "trace")]
+            let task = task.instrument(task_span);
+            scope.spawn_on_scope(task);
```

Comment:
You need this because the executor is allowed to run on any thread and exclusive systems can access non send resources. If we remove those from the world then we could do these optimizations. It would also be possible if we end the scope somehow before running the exclusive systems, but that would be a significantly different executor design.

Reply:
Exactly.